core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala

```diff
@@ -96,6 +109,47 @@ private[spark] class ApplicationInfo(
 
   private[master] def coresLeft: Int = requestedCores - coresGranted
 
+  /**
+   * Return the number of executors waiting to be scheduled once space frees up.
+   *
+   * This is only defined if the application explicitly set the executor limit. For instance,
+   * if an application asks for 8 executors but there is only space for 5, then there will be
+   * 3 waiting executors.
+   */
+  private[master] def numWaitingExecutors: Int = {
+    if (executorLimit != Integer.MAX_VALUE) {
+      math.max(0, executorLimit - executors.size)
+    } else {
+      0
+    }
+  }
+
+  /**
+   * Add a worker to the blacklist, called when the executor running on the worker is killed.
+   * This is used only if cores per executor is not set.
+   */
+  private[master] def blacklistWorker(workerId: String): Unit = {
+    blacklistedWorkers += workerId
+  }
+
+  /**
+   * Remove workers from the blacklist, called when the application requests new executors.
+   * This is used only if cores per executor is not set.
+   */
+  private[master] def removeFromBlacklist(numWorkers: Int): Unit = {
+    blacklistedWorkers.take(numWorkers).foreach { workerId =>
```

Comment:
Still, there must be a more direct way. Doing a `foreach` for this seems really an anti-pattern. This looks clearer:

```
blacklistedWorkers --= blacklistedWorkers.take(numWorkers)
```